pkg/vsphere/disk/disk_manager.go

```diff
@@ -161,20 +161,25 @@ func (m *Manager) CreateAndAttach(op trace.Operation, config *VirtualDiskConfig)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	d.l.Lock()
+	defer d.l.Unlock()
 
 	blockDev, err := waitForDevice(op, devicePath)
 	if err != nil {
-		op.Errorf("waitForDevice failed for %s with %s", config.DatastoreURI, errors.ErrorStack(err))
+		op.Errorf("waitForDevice failed for %s with %s", d.DatastoreURI, errors.ErrorStack(err))
 		// ensure that the disk is detached if it's the publish that's failed
 
-		if detachErr := m.Detach(op, config); detachErr != nil {
-			op.Debugf("detach(%s) failed with %s", config.DatastoreURI, errors.ErrorStack(detachErr))
+		disk, findErr := findDiskByFilename(op, m.vm, d.DatastoreURI.String())
+		if findErr != nil {
+			op.Debugf("findDiskByFilename(%s) failed with %s", d.DatastoreURI, errors.ErrorStack(findErr))
+		}
+
+		if detachErr := m.detach(op, disk); detachErr != nil {
+			op.Debugf("detach(%s) failed with %s", d.DatastoreURI, errors.ErrorStack(detachErr))
 		}
 
 		return nil, errors.Trace(err)
 	}
-
-	d.ParentDatastoreURI = config.ParentDatastoreURI
 	d.setAttached(blockDev)
```

Comment:
CreateAndAttach is WaitTillDiskAppearsInTheGuest and Attach is WaitTillReconfigure finishes, I'm guessing that's why it is separated that way (coming from the original code)